app/lib/admin/backend.dart

```diff
@@ -271,6 +277,117 @@ class AdminBackend {
     });
   }
 
+  Future<void> removePackage(String packageName) async {
+    final caller = await _requireAdminPermission(AdminPermission.removePackage);
+
+    _logger.info('${caller.userId} (${caller.email}) initiated the delete '
+        'of package $packageName');
+
+    List<Version> versionsNames;
+    await withRetryTransaction(_db, (tx) async {
+      final deletes = <Key>[];
+      final Key packageKey = _db.emptyKey.append(Package, id: packageName);
+      final package = (await tx.lookup([packageKey])).first as Package;
+      if (package == null) {
+        _logger
+            .info('Package $packageName not found. Removing related elements.');
+      } else {
+        deletes.add(packageKey);
+      }
+
+      final versionsQuery = tx.query<PackageVersion>(packageKey);
+      final versions = await versionsQuery.run().toList();
+      versionsNames = versions.map((v) => v.semanticVersion).toList();
+      deletes.addAll(versions.map((v) => v.key));
+
+      final moderatedPkgKey =
+          dbService.emptyKey.append(ModeratedPackage, id: packageName);
+      ModeratedPackage moderatedPkg = await dbService
+          .lookupValue<ModeratedPackage>(moderatedPkgKey, orElse: () => null);
+      if (moderatedPkg == null) {
+        moderatedPkg = ModeratedPackage()
+          ..parentKey = _db.emptyKey
+          ..id = packageName
+          ..name = packageName
+          ..moderated = DateTime.now().toUtc()
+          ..versions = versions.map((v) => v.version).toList();
+
+        if (package != null) {
```

Comment:
If the `package == null`, should we even create the `ModeratedPackage` instance? I guess it has a use for backfilling previously deleted packages, but otherwise it feels like we shouldn't create one.

Reply:
I'm not entirely sure. I did consider it and I was thinking, if we initiate a package removal for whatever reason, we would probably want the name to be reserved, no? And in that case we need to create a `ModeratedPackage` instance.

Reply:
sgtm.